nixedit: init at 1.0.0

fndov opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
A NixOS Multipurpose CLI/TUI Utility. https://github.com/fndov/nixedit
#### Features:
- Full Terminal User Interface & Command line usage.
- Search Configure Build Backup Update Delete Optimise in one step.
- Complete Profile management, Integrated Creation & Removal. 
- Integrates with Github to upload backups of your configuration.

I'm sure there are no problems/bugs with the package, the last step was modifying default.nix. 👍
## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Commit:
nixedit: init at 1.0.0

Commit:
nixfmt

Commit:
format should work now

Commit:
installCheckPhase into /dev/null

drupol commented on pkgs/tools/nix/nixedit/default.nix:
```diff
@@ -0,0 +1,60 @@
+{
+  pkgs ? import <nixpkgs> { },
```
Don't pass `pkgs`, see an example: https://github.com/NixOS/nixpkgs/pull/348180/files

drupol commented on pkgs/tools/nix/nixedit/default.nix:
```diff
@@ -0,0 +1,60 @@
+  src = pkgs.fetchFromGitHub {
+    owner = "fndov";
+    repo = "nixedit";
+    rev = "main";
```
`main` is a branch, that means that building this derivation next time, it might fail since the hash will be invalid. You need to fix the src using a commit ID, see an example at https://github.com/NixOS/nixpkgs/pull/348180/files

drupol commented on pkgs/tools/nix/nixedit/default.nix:
```diff
@@ -0,0 +1,60 @@
+    owner = "fndov";
+    repo = "nixedit";
+    rev = "main";
+    sha256 = "1hmda40xcqdb5akbbafyhhkhnarn677fjvd264ki5pbz71lmlbfl";
```
Prefer `hash = ...` to `sha256 = ...` Both currently work, but the former will make it easier if the algorithm ever needs to change.

Since there could be different hashing algorithms in use, this make sense.

More info: https://nixos.org/manual/nixpkgs/stable/#fetchurl

> Nixpkgs contributors are currently recommended to use hash.

drupol commented on pkgs/tools/nix/nixedit/default.nix:
```diff
@@ -0,0 +1,60 @@
+  };
+
+  nativeBuildInputs = [
+    pkgs.makeWrapper
```
Remove the `pkgs.` prefix everywhere and pass them as top level arguments.

drupol commented on pkgs/tools/nix/nixedit/default.nix:
```diff
@@ -0,0 +1,60 @@
+    $out/bin/nixedit --help > /dev/null
+  '';
+
+  meta = with pkgs.lib; {
```
Missing `meta.mainProgram`.

fndov reviewed (commented):
Looks good

fndov reviewed (commented):
Thanks for the suggestion, I'll look into this 👍

Commit:
Solve every request: Done.

I tested the updated default.nix inside QEMU and used nixfmt before committing this. Hope this is everything left, let me know if there's anything I messed. The example links were helpful thanks 👍

drupol commented on pkgs/tools/nix/nixedit/default.nix:
```diff
@@ -0,0 +1,73 @@
+
+    # Wrap nixedit to include the necessary dependencies in PATH
+    wrapProgram $out/bin/nixedit --prefix PATH : \
+      "${bash}/bin:${coreutils}/bin:${nix-tree}/bin:${jq}/bin:${micro}/bin:${git}/bin:${fzf}/bin:${dialog}/bin"
```
Please use makeBinPath.

Commit:
Use makeBinPath

drupol commented on pkgs/tools/nix/nixedit/default.nix:
```diff
@@ -0,0 +1,84 @@
+    # Wrap nixedit to include the necessary dependencies in PATH
+    wrapProgram $out/bin/nixedit --prefix PATH : \
+      "${
+        lib.makeBinPath [
```
If you use the `finalAttrs` pattern, you could simply reuse the list defined in `buildInputs` and avoid duplicates.

drupol replied:
I think we should use the `finalAttrs` pattern here.

The `finalAttrs` pattern will let you remove the `rec` keyword ([see its implementation in Nix](https://github.com/NixOS/nix/blob/2d1d81114d72ace89ce08cd3bc93f4eb27a2975d/src/libexpr/nixexpr.cc#L121)).

Why is this pattern preferred to `rec` ?

Let's take this simple code example:

```nix
mkDerivation rec {
   foo = 1;
   bar = foo + 1;
}
```

and then `.overrideAttrs(old: { foo = 2; })`, you'll get `{ foo = 2; bar = 2; }` while with `finalAttrs` pattern, it would work correctly because it's a real fixed point.

Let me share a couple of useful links regarding the `finalAttrs` pattern:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

Feel free to reach out if you need some assistance.

Commit:
Reuse buildInputs instead of repeating

drupol commented on pkgs/tools/nix/nixedit/default.nix:
```diff
@@ -13,15 +13,15 @@
 }:
 
-stdenv.mkDerivation {
+stdenv.mkDerivation rec {
```
I think we should use the `finalAttrs` pattern here.

The `finalAttrs` pattern will let you remove the `rec` keyword ([see its implementation in Nix](https://github.com/NixOS/nix/blob/2d1d81114d72ace89ce08cd3bc93f4eb27a2975d/src/libexpr/nixexpr.cc#L121)).

Why is this pattern preferred to `rec` ?

Let's take this simple code example:

```nix
mkDerivation rec {
   foo = 1;
   bar = foo + 1;
}
```

and then `.overrideAttrs(old: { foo = 2; })`, you'll get `{ foo = 2; bar = 2; }` while with `finalAttrs` pattern, it would work correctly because it's a real fixed point.

Let me share a couple of useful links regarding the `finalAttrs` pattern:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

Feel free to reach out if you need some assistance.

Commit:
Replace rec with finalAtters

fndov commented:
I'm currently working on the stable branch, should be finished very soon. Then default.nix won't use a hard codded version, making updates simple.

Commit:
Finished Stable branch, rec problem is solved to my knowledge.

Commit:
Use commit ID. package is ready.

Commit:
Syntax improvment. Merge ready.

Commit:
nixfmt.

drupol reviewed (changes requested):
Please move it under pkgs/by-name , also make sure this PR only contains only one commit.

fndov commented:
I will create a new PR. With the pkgs/by-name change. And link it here.

fndov commented:
https://github.com/NixOS/nixpkgs/pull/349119 @drupol
